src/commands/giveaway.rs

```diff
@@ -0,0 +1,650 @@
+
+use serenity::{
+    builder::{CreateComponents, CreateEmbed, CreateEmbedFooter},
+    model::{
+        interactions::{
+            message_component::ButtonStyle, InteractionApplicationCommandCallbackDataFlags,
+            InteractionResponseType,
+        },
+        prelude::application_command::*,
+    },
+    prelude::*,
+};
+
+use crate::{
+    database::Database, extensions::*, models::Giveaway, Http, Interaction, Message, ReactionType,
+    User,
+};
+
+static DEFAULT_DURATION: i64 = 3600;
+static DEFAULT_WINNERS: i64 = 1;
+static DEFAULT_PRIZE: &str = "Nothing :(";
+pub static GIVEAWAY_REACTION_CHAR: char = '🎉';
+
+struct ListOffset;
+
+impl TypeMapKey for ListOffset {
+    type Value = HashMap<u64, i64>;
+}
+
+async fn ensure_offset_map(ctx: &Context) {
+    let mut data = ctx.data.write().await;
+    if !data.contains_key::<ListOffset>() {
+        data.insert::<ListOffset>(HashMap::new());
+    }
+}
+
+fn footer_with_text<S: Into<String>>(text: S) -> CreateEmbedFooter {
+    let mut footer = CreateEmbedFooter {
+        0: std::collections::HashMap::new(),
+    };
+    footer.text(text.into());
+    footer
+}
+
+fn generate_list_components(offset: i64, giveaways: i64) -> CreateComponents {
+    let mut c = CreateComponents { 0: Vec::new() };
+
+    c.create_action_row(|r| {
+        r.create_button(|b| {
+            b.style(ButtonStyle::Secondary);
+            b.custom_id("GIVEAWAY_list_back");
+            b.label("Previous page");
+            b.disabled(offset - 10 < 0)
+        });
+        r.create_button(|b| {
+            b.style(ButtonStyle::Secondary);
+            b.custom_id("GIVEAWAY_list_next");
+            b.label("Next page");
+            b.disabled(offset + 10 >= giveaways)
+        })
+    });
+    c
+}
+
+async fn generate_list_embeds(db: &Database, offset: i64) -> Vec<CreateEmbed> {
+    let giveaways = db.get_n_giveaways_with_offset(10, offset).await.unwrap();
+    let mut giveaway_winners = Vec::with_capacity(giveaways.len());
+
+    for g in giveaways.iter() {
+        let winners = db.get_giveaway_winners(g.id).await.unwrap();
+        giveaway_winners.push(winners);
+    }
+
+    let mut embeds: Vec<CreateEmbed> = Vec::new();
+    for (g, winners) in giveaways.iter().zip(giveaway_winners.iter()) {
+        let mut e = CreateEmbed { 0: HashMap::new() };
+        let winner_string = winners
+            .iter()
+            .map(|x| format!("<@{}>", x.user_id.to_string()))
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        e.title(format!("Giveaway #{}", g.id));
+        e.description(format!(
+            "**Prize**: {}\n**Winners**: {}\n**End time**: <t:{}:R>",
+            g.prize,
+            if g.completed {
+                winner_string
+            } else {
+                format!("Max {}", g.max_winners)
+            },
+            g.end_time.timestamp()
+        ));
+        embeds.push(e);
+    }
+    embeds
+}
+
+async fn get_reacters(
```

Comment:
This, I believe is a misspelling of "reactor"

Reply:
If I understood https://wikidiff.com/reacter/reactor correctly "reacter" would be the correct term to use here

Reply:
both words most like fit the definition then: <https://www.merriam-webster.com/dictionary/reactor>
mb. resolved.